### DISCUSSION TITLE
Prompt / deduce root for nested project?

### DISCUSSION THREAD

**Discussion by mmgeorge** [emacs-lsp/lsp-mode#4301] [Q&A] [answered]
I have a project structure that consists of multiple nested projects with different languages. At the root of the project I have a rust project, and then a shaders sub-project that uses a very WIP WGSL language server that I am currently working on:

my-project
- Cargo.toml
- src
  - lib.rs
- shaders 
  - wgslx.toml <- shader project definition file 
  - src
    - lib.wgslx

For my lsp client, I have simply: 
```
(lsp-register-client
 (make-lsp-client :new-connection (lsp-stdio-connection "<my-local-path>/wgslx-lsp")
                  :activation-fn (lsp-activate-on "wgsl")
                  :server-id 'wgslx))
```

and I have the projectile project type registered with: 

```
(when (require 'projectile nil 'noerror)
  (projectile-register-project-type 'wgslx '("wgslx.toml")
                                    :project-file "wgslx.toml"
                                    :src-dir "src"))
```

Is there a way to make it so that I can infer different language server roots for these different projects? If I first goto src/lib.rs, I am prompted to set the root which I do at my-project/. But when I navigate to my-project/shaders/src/lib.rs, I don't get prompted to set the root for the sub-project.

Maybe I'm not approaching this the right way, and should instead try to make the server more robust? E.g.,  given a root at my-project/ and a first WGSL file request at my-project/shaders/src/lib.wgslx, walk up until I find my wgsl project file & manage that internally as the root for the WGSL language server?

Thanks!

**Comment by yyoncho** [comment, marked as answer]
You can either manually use M-x lsp-workspace-folders-add or turn on auto guess project root (and hope for the best) by setting lsp-auto-guess-root to t.

**Reply by mmgeorge** [reply]
No luck with `lsp-auto-guess-root`, but `lsp-workspace-folders-add` works after I restarted the server. Thanks for the quick response!
